### DISCUSSION TITLE
Ecosystem: What features would you like to see?

### DISCUSSION THREAD

**Discussion by dataders** [dbt-labs/dbt-core#6104] [Community]
> “The most dangerous worldview is the worldview of those have not viewed the world.”
> ― Alexander von Humboldt

Note: This focus of this discussion is for folks building tools that integrate with dbt-core, not the entire dbt community at large. Perhaps we should start a topic just for that? 

Friends in the dataverse ecosystem. In what areas could dbt-core to help better integrate with your product and make it successful?

Stable Python API? Better adapter unit tests? SQL Parser? Jinja3?

I'll post some answers I've heard from the communtiy. Please upvote and comment with more info.

If you don't feel comfortable broadcasting your hopes/dreams/desires, feel free to reach out via private channels.

**Comment by amirbtb** [comment]
My Christmas wishlist :
- Use dbt as a Python library
- being able to define business entities in YAML syntax so that dbt can generate SQL with joins
- column level lineage and all it can bring
- codegen : generate source YAML for external tables
- Mainstream tools (Power BI, Tableau, Looker Studio,...) integrating semantic layer
- open source alternatives for the dbt proxy + dbt Metadata API (on top of a Git repo)
- dbt support for both batch & streaming

**Comment by dataders** [comment]
Cross project lineage (https://github.com/dbt-labs/dbt-core/discussions/5244).

**Comment by dataders** [comment]
Multi-data platform projects. This is the converse of cross-project lineage. Rather than make it easy to stitch multiple dbt projects together, this is about, can I have a single dbt project with one DAG, but where `model_B.sql` in database B can `ref` a model that lives in database A.

This is similar to https://github.com/dbt-labs/dbt-core/discussions/5073, but different in that in this scenario there is no internal/external distinction, rather that everything becomes "internal"

**Reply by nicku33** [reply]
This would be great: at MotherDuck we'd love it for getting duckdb into the mix of existing models, but it implies tuple transport across databases via external storage. Is it within the future design scope of dbt to get into that ? Some db's are pretty good at parquet in and out of S3 without much setup.

**Reply by jtcohen6** [reply]
There's some preliminary conversation / ideation about this over in https://github.com/dbt-labs/dbt-core/discussions/5758#discussioncomment-3630741

**Comment by dataders** [comment]
more standardized Python API. Today, we strongly recommend that folks don't take a dependency on our internal API. But people do anyway. Folks access the `Manifest` class instead of the `manifest.json`. I'll add more examples later.

Some ecosystem folks have asked for this.

**Reply by lostmygithubaccount** [reply]
on this one -- tagging @amirbtb, @ma1ster as well -- what are you looking to use a dbt Python API/SDK for? one of our worries would be users defining the entire project/DAG within the Python SDK. are you envisioning this for programmatically interacting with the project after it's already built, or defining models programmatically via Python?

**Reply by amirbtb** [reply]
Today I am running dbt in a Google Cloud Run & Cloud Sceduler. I am calling the dbt CLI using a Python subprocess, but I would prefer to run dbt commands directly using a Python library.
I know there is a way to do it since dbt is coded in Python but since you recommend to not use it & it is poorly documented, I've always avoided it.

**Comment by dataders** [comment]
Actual unit testing. i.e. testing the transformation code not the data output. One component of this is running tests against static data such as csvs and jsons
https://github.com/dbt-labs/dbt-core/discussions/4455

**Comment by ma1ster** [comment]
Definitely using dbt as python library

**Comment by wooddar** [comment]
Hugo from Hightouch here! Will reply to my own thread if anything else comes up but the top of mind right now is definitely:
* Exposures for Analyses - Our customers are heavily using dbt analyses in their syncs and exposures created referring to analyses are 'broken'
* Other people have mentioned column-level lineage but I would specifically love dbt support for 'schema discovery' where *all* model columns can be added to the manifest e.g. after `dbt compile` and not just the ones with metadata.

**Reply by jtcohen6** [reply]
@wooddar Thanks for weighing in!

> Our customers are heavily using dbt analyses in their syncs and exposures created referring to analyses are 'broken'

Is the idea here, being able to "save" a query that gets passed into the exposure? Could you imagine wanting to associate an exposure with a SQL query / `.sql` file, rather than the indirection of `analysis --> exposure`?

> dbt support for 'schema discovery' where all model columns can be added to the manifest e.g. after dbt compile and not just the ones with metadata.

This information is available from `catalog.json` (produced by `docs generate`) today. Is that a no-go? What if a model's full set of columns were available after `dbt run` (still requires materializing the model in the database)?

**Reply by wooddar** [reply]
> Is the idea here, being able to "save" a query that gets passed into the exposure? Could you imagine wanting to associate an exposure with a SQL query / .sql file, rather than the indirection of analysis --> exposure?

Relating an exposure to some generic SQL file would also be fine as long as it appears appropriately on our customers' lineage graphs once our exposures are generated?

> This information is available from catalog.json (produced by docs generate) today. Is that a no-go? What if a model's full set of columns were available after dbt run (still requires materializing the model in the database)?

Interesting, I suppose we could use this though if our customers are using analyses then we would still need to manually inspect their schema...

**Comment by morsapaes** [comment]
From our side: +1 on cross-project lineage and multi-adapter runs, since it'll be somewhat common for users to deploy Materialize alongside a batch data warehouse! In addition to that, there are a couple of features that would be nice to have:

### Multiplexed relations (aka one command = multiple relations)

For specific source types in Materialize, a single `CREATE SOURCE` command will create a _source_ plus a varying number of _sub-sources_, leading to something like the following:

```
      "pg_source"
       /       \
depends_on   depends_on
     /           \
    v             v
 "foo"            "foo2"
```

There seems to be no easy way to pull these sub-sources into the dbt context. The current workflow is to manually create a bunch of `.yml` and `.sql` files to piece things together, but in real-world scenarios a single source might have 100+ sub-sources. That's a lot of files!

Existing issues and discussions we've looked into in the past that seemed promising, but didn't seem to cut it: [#5100](https://github.com/dbt-labs/dbt-core/discussions/5100), [#5101](https://github.com/dbt-labs/dbt-core/discussions/5101), [#2716](https://github.com/dbt-labs/dbt-core/issues/2716), just hacking the [`graph`](https://docs.getdbt.com/reference/dbt-jinja-functions/graph) context variable (👹).

### Runnable exposures

In the same way that it can ingest data from external systems, Materialize can also sink data to external systems. Exposures sound like a great way to model this (with a side of external nodes?), but the initial implementation has a few caveats for our specific needs: 1) it isn't possible to use a custom string as the exposure type; 2) exposures aren't compilable or runnable, and a sink requires that we ship a `CREATE SINK` statement to Materialize.

We'd just **really** love to not see any raw DDL statements in a `dbt-materialize` project, but can definitely live with the current state of having custom `source` and `sink` materializations! We wrote [some musings](https://gist.github.com/morsapaes/03cb046ba4a984da644263c7e22cf094) around this a while back, and explored some of the proposed options with no luck!

**Reply by jtcohen6** [reply]
Yes & yes!

As far as a [guide for the multiplexed](https://en.wikipedia.org/wiki/The_Guide_for_the_Perplexed), I have one more discussion to throw your way: https://github.com/dbt-labs/dbt-core/discussions/6185 (**update:** I previously linked to #5686 and #6067, which could still be related older reading)

As exposures go, it is simply so clear to me that they ought to be our way of modeling syncs & sinks, and that they need to actually do a thing. That thing might be SQL (`create sink`) in Materialize, or a webhook / API call in other cases (https://github.com/dbt-labs/dbt-core/discussions/5073 - you know what link this is, you don't even need to hover).

**Reply by morsapaes** [reply]
Thanks, @jtcohen6! I'll hop on #6185 for comments specific to multiplexed sources. And great that we're aligned when it comes to exposures!

While reading through the helpful links you leave in your trail, I realized that I'd missed #5686. The ideas in there line up really well with an issue we have out for [dynamic unions](https://github.com/MaterializeInc/materialize/issues/13743). In the context of dbt, this would be useful to handle the (possible disparate amount of) materialized views that are created as a result of our take on `--store-failures` (_aka_ continuous testing).

**Comment by hovaesco** [comment]
Hi @dataders 

It would be nice to have improved test coverage for `dbt-tests-adapter`, like to have ConnectionManager integration testing without the test having to invoke dbt.

**Comment by joshuataylor** [comment]
I would love to see column types be settable in the YAML, so you can say "this is a number(38,0) DO NOT CHANGE IT". I hate hardcoding the column types in SQL like `something::number(38,0) as something` (it's a pita).

**Reply by sungchun12** [reply]
Well, would you look at that. We're working on that as we speak :) https://github.com/dbt-labs/dbt-core/issues/6079

**Reply by joshuataylor** [reply]
WOOOOOOOOO

I'm happy to help contribute and test the Snowflake portion.

**Comment by cpstein** [comment]
On the Hightouch side, we have an important request for a bugfix!

The issue is detailed in: https://github.com/dbt-labs/dbt-snowflake/issues/112.

As flagged in the Github issue, `truncate` and `insert` commands currently run in different transactions. This is particularly problematic for Hightouch users who use dbt (esp with frequent syncs like every minute) as there is a chance that Hightouch will try to sync after the table has been truncated, but before data has been inserted into it, causing Hightouch to not sync the intended data correctly (including syncing data that should not be synced to a destination!).

**Reply by jtcohen6** [reply]
@cpstein Thanks for flagging! That issue should only be affecting tables produced by `dbt seed` (CSV uploads). Is it common for dbt + Hightouch users to schedule syncs of data produced from seeds? That's not something I would have expected, though it is technically possible

**Reply by cpstein** [reply]
@jtcohen6 Probably not the most common use case, but we have seen a couple cases of this.

**Comment by virajmparekh** [comment]
Hey all! From Astronomer's perspective:

- I think the biggest thing we'd like to see is a a first class python library. A lot of our customers have to dance around breaking changes with python dependencies in their Airflow DAGs to get things to work, and many have even opted to run each dbt model in its own container to get around this. I don't know if this is possible, but it'd be cool if a single PyPI packaged allowed users to call the cloud api as well as use core. 

- Additionally, we'd love an interface to getting lineage events from the core package. We have users that wrap the core package and run a modified command to consume those events - a first class way to support that would be awesome.

Love the public nature of the discussion here!
